samza-core/src/main/java/org/apache/samza/table/caching/CachingTable.java

```diff
@@ -105,96 +102,222 @@ public void init(SamzaContainerContext containerContext, TaskContext taskContext
+                .collect(Collectors.toList()));
+            getAllResult.putAll(records);
+          }
+          readMetrics.getAllNs.update(System.nanoTime() - startNs);
+          return getAllResult;
+        }
+      });
   }
 
   @Override
   public void put(K key, V value) {
+    try {
+      putAsync(key, value).get();
+    } catch (InterruptedException e) {
+      throw new SamzaException(e);
+    } catch (Exception e) {
+      throw (SamzaException) e.getCause();
+    }
+  }
+
+  @Override
+  public CompletableFuture<Void> putAsync(K key, V value) {
     writeMetrics.numPuts.inc();
-    long startNs = System.nanoTime();
     Preconditions.checkNotNull(rwTable, "Cannot write to a read-only table: " + rdTable);
-    Lock lock = stripedLocks.get(key);
+
+    long startNs = System.nanoTime();
+    return rwTable.putAsync(key, value).handle((result, e) -> {
+        if (e != null) {
+          throw new SamzaException(String.format("Failed to put a record, key=%s, value=%s", key, value), e);
+        } else if (!isWriteAround) {
+          if (value == null) {
+            cache.delete(key);
+          } else {
+            cache.put(key, value);
+          }
+        }
+        writeMetrics.putNs.update(System.nanoTime() - startNs);
+        return result;
+      });
+  }
+
+  @Override
+  public void putAll(List<Entry<K, V>> records) {
     try {
-      lock.lock();
-      rwTable.put(key, value);
-      if (!isWriteAround) {
-        cache.put(key, value);
-      }
-    } finally {
-      lock.unlock();
+      putAllAsync(records).get();
+    } catch (InterruptedException e) {
+      throw new SamzaException(e);
+    } catch (Exception e) {
+      throw (SamzaException) e.getCause();
     }
-    writeMetrics.putNs.update(System.nanoTime() - startNs);
   }
 
   @Override
-  public void putAll(List<Entry<K, V>> entries) {
+  public CompletableFuture<Void> putAllAsync(List<Entry<K, V>> records) {
     writeMetrics.numPutAlls.inc();
     long startNs = System.nanoTime();
     Preconditions.checkNotNull(rwTable, "Cannot write to a read-only table: " + rdTable);
-    entries.forEach(e -> put(e.getKey(), e.getValue()));
-    writeMetrics.putAllNs.update(System.nanoTime() - startNs);
+    return rwTable.putAllAsync(records).handle((result, e) -> {
+        if (e != null) {
+          throw new SamzaException("Failed to put records " + records, e);
+        } else if (!isWriteAround) {
+          Set<K> allKeys = new HashSet<>();
+          List<K> delKeys = new ArrayList<>();
+          List<Entry<K, V>> putRecords = new ArrayList<>();
+          // Separate out put vs delete records
+          records.forEach(r -> {
+              allKeys.add(r.getKey());
+              if (r.getValue() != null) {
+                putRecords.add(r);
+              } else {
+                delKeys.add(r.getKey());
+              }
+            });
+
+          cache.deleteAll(delKeys);
+          cache.putAll(putRecords);
```

Comment:
Can cache.putAll with null-key automatically delete the key from the cache? it avoids having to bifurcate puts and deletes. The putAllAsync on the "table" seems to be following that semantic anyway. Since, a "cache" is also implemented as a "table", should we be consistent?

Reply:
Good point. I'll update the guava table cache implementation to handle null-key put case.